src/systems/buoyancy_engine/BuoyancyEngine.cc

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <algorithm>
+#include <chrono>
+#include <memory>
+#include <mutex>
+#include <string>
+
+#include <ignition/gazebo/Link.hh>
+#include <ignition/msgs/double.pb.h>
+#include <ignition/plugin/Register.hh>
+#include <ignition/transport/Node.hh>
+
+#include "BuoyancyEngine.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+class ignition::gazebo::systems::BuoyancyEnginePrivateData
+{
+  /// \brief Callback for incoming commands
+  /// \param[in] volumeSetPoint - ignition message containing the desired volume
+  /// (in cc) to fill/drain bladder to.
+  public: void OnCmdBuoyancyEngine(
+    const ignition::msgs::Double& volumeSetPoint);
+
+  /// \brief current volume of bladder in m^3
+  public: double bladderVolume = 3e-5;
+
+  /// \brief Maximum inflation rate in m^3*s^-1
+  public: double maxInflationRate = 3e-6;
+
+  /// \brief Set-point for volume
+  public: double volumeSetPoint = 0.000030;
+
+  /// \brief Minimum volume of bladder in m^3
+  public: double minVolume = 0.000030;
+
+  /// \brief Maximum volume of bladder in m^3
+  public: double maxVolume = 0.000990;
+
+  /// \brief The link which the bladder is attached to
+  public: ignition::gazebo::Entity linkEntity;
+
+  /// \brief The gravitational force kg*m*s^-2
+  public: double gravity = 9.81;
+
+  /// \brief The fluid density in kg*m^-3
+  public: double fluidDensity = 1000;
+
+  /// \brief The neutral volume in m^3
+  public: double neutralVolume = 0.0003;
+
+  /// \brief Trasport node for control
+  public: ignition::transport::Node node;
+
+  /// \brief Get bladder status
+  public: ignition::transport::Node::Publisher statusPub;
+
+  /// \brief mutex
```

Comment:
It's helpful to document what it is meant to protect

Reply:
Resolved in 8ba2e0e